Introduce new `tag-to-name` option

daniil-berg opened the pull request:
I like this action a lot, but I would like to have the option for it to automatically use the **entire tag message** as the release text body and use the **tag itself** (e.g. `v1.0.42`) as the name/title of the release. I hope that makes sense.

I know I could use your action as is and just repeat the tag name in the subject line of the message, but this would be a redundancy that I do not like very much.

I could also just manually edit the release text and title after it was auto-generated by the action, but this partially defeats the purpose of using an automated generation process.

These changes should keep your current approach of using the git tag subject line as the release name and the rest of the message as the body as the default, so it should not break anything.

My experience with TypeScript is almost non-existent, so I hope my changes are not completely broken. If you think this PR is worthwhile and something is wrong/missing, I'll do my best to fix it.

spenserblack commented on action.yml:
```diff
@@ -17,6 +17,15 @@ inputs:
     description: If the release should be a draft
     required: true
     default: false
+  tag-as-name:
```
:thinking: I'm going to brainstorm and see if there's a more descriptive name. Can't think of one now, but I wonder if there's an input name that can capture both
1. the tag's name becomes the title
2. the *entire* tag's contents becomes the body

But, at least, I think that `title` is more accurate than `name`. If you use `gh release create`, for example, you're prompted for the release's *title.*

daniil-berg replied:
You are right. I used the term `title` everywhere at first, but was discouraged, when I saw that the [Github REST API](https://docs.github.com/en/rest/releases/releases?apiVersion=2022-11-28#create-a-release) defines the field for release creation as `name`. So I changed everything to `name`. I guess it is an inconsistency on Github's part. If you like `title` more as well, I am all for it. But then you may want to use it throughout the existing code as well instead of `name`.

As for the option name, I doubt it makes sense to try and convey _both_ in it. I think `tag-as-title` indicates the first point, which is what I initially wanted to happen. Alternatively, we could simply call that option `title` and allow it to be passed either `tag` or `subject` as arguments for now. Basically an enum. That would allow adding more choices later on. Maybe something like `both` or `tag-and-subject` to result in something like `v1.0.1 (This is the subject line)` or something to that effect. But that is just me brainstorming as well.

daniil-berg replied:
Or how about this:

Introduce the `title` parameter that accepts _any_ string. But we allow special placeholders for the tag name or the subject line in it. The user can specify something like `'$tag ($subject)'` and would get `v1.0.1 (This is the subject line)` as the title. That gives maximum flexibility.

And in addition we have the option `text` with the enum `empty`, `subject`, `body`, `contents` or `contents-signed` (or something similar). Maybe overkill with that last option...

spenserblack replied:
:thinking: After thinking about it, `tag-as-title` is probably best.

> Introduce the `title` parameter that accepts _any_ string. But we allow special placeholders for the tag name or the subject line in it. The user can specify something like `'$tag ($subject)'` and would get `v1.0.1 (This is the subject line)` as the title. That gives maximum flexibility.

That's a really fun idea, but, for now, I think let's just stick with `tag-as-title: true|false`. That level of complexity is kind of what this action is trying to abstract away -- there isn't much different between `$subject` and `(contents:subject)` IMO. A sort of templating input probably belongs in a different action IMO.

spenserblack reviewed (commented):
Thanks for the PR! I left some feedback, let me know what you think.

Additionally, as the first potential user of this feature, what do you think about this:
As you know, the tag's first line is the subject. For this reason, would it make sense to prepend `# ` to the body, so that the tag's subject becomes a level 1 header for the release notes?

daniil-berg commented on src/tag.ts:
```diff
@@ -4,11 +4,11 @@ export default class Tag {
+    return await this.getTagContents(includeSubject ? 'contents' : 'contents:body')
   }
 
   private async getTagContents(contents: string): Promise<string> {
```
Great idea. I'll implement that once you decide on the correct terminology (`name` vs. `title`).

daniil-berg replied:
I just remembered something that I forgot in my initial implementation of those changes.

When you just use `%(contents)` and your tag is GPG-signed, it includes that `-----BEGIN PGP SIGNATURE-----` block at the bottom as well.

And as far as I know there is no filter for just that built into git. I solved it like this for myself:

```bash
git tag -v --format='%(contents)' TAG_NAME | awk '/^-*BEGIN PGP SIGNATURE-*$/ {exit} {print}'
```

This gives me the entire thing minus the signature.

This of course also works, if that block is missing, so adding that to the actual `git` call should work in any case.

spenserblack replied:
Good catch! I'll spend some time researching different ways to filter out the signature.

spenserblack replied:
OK, so a couple of ideas.

## Bad Ideas

- `--format=%(contents:lines=1000)` will include the subject and body, but exclude the signature
- `--format=%(contents:signature)` will get the signature, and you can use `String.replace`

## Workable Idea

```shell
git tag -v --format='%(contents:subject)

%(contents:body)' $TAG_NAME
```
This will basically rebuild the tag message. Sadly, the code won't look as clean and pretty :laughing:

******

Here's what I'm thinking:
```typescript
// Let's just require a filter for now...
getContents(filter: string): string;
getContents(filter: string[]): string[];
getContents(filter: string | string[]): string | string[] {
  if (typeof filter === 'string') {
    // str = ...
    return str;
  }
  return filter.map((s: string) => this.getContents(s));
}
```

See [function overloads](https://www.typescriptlang.org/docs/handbook/2/functions.html#function-overloads), and let me know if you have any questions.

Then, when `tag-as-title` is `true`, you call `getContents(['subject', 'body'])`, and join with newlines.

daniil-berg replied:
I like it, but I am having trouble implementing the overloaded function. I tried out the codespace you suggested, but the linter keeps telling me about `getContents` being a duplicate name. :confused:  I must be doing something wrong. If you don't mind making these adjustments yourself and pushing them into this branch, that would probably be more efficient. I am on board with all your suggested changes.

spenserblack replied:
That works for me. I'll get to it when I have time.
I've never *formally* learned Typescript, and [sometimes the typing system baffles me](https://github.com/jamiebuilds/json-parser-in-typescript-very-bad-idea-please-dont-use), so I may have given incorrect instructions :shrug:

spenserblack replied:
@daniil-berg You weren't doing anything wrong, it was a [false catch](https://github.com/typescript-eslint/typescript-eslint/issues/291).

https://github.com/spenserblack/actions-tag-to-release/pull/114/commits/5412c9b2a9c38d26650d4cfc9b0d106ac2d78391

spenserblack commented:
> Does that make sense?

Yes, sounds good! :+1:

spenserblack commented:
@daniil-berg Pushed up some changes, want to try them out?

daniil-berg commented:
I added some more changes. The `tag` contents filter was a typo I think because it is not valid. You meant to use `subject`.

Also as you said `title` is a more suitable term than `name` in this context, I changed it everywhere.

Also changed/refactored the `Tag`  methods and their calls a bit. Hope this is OK.

I guess my build step did not work for some reason. Not sure... If the rest is fine with you, you can finalize it. Thanks for considering this.

spenserblack commented:
Just rebuilt `dist/`.

Would you be able to run `daniil-berg/actions-tag-to-release@tag-as-title` and confirm if you get the expected behavior?

daniil-berg commented:
Hm, just ran it with `tag-as-title: true`, but now I am getting a strange result, where the title shown for the release is `tag: subject` (e.g. `v1.2.3: this is text`.

What is even stranger though is when I go to manually edit the release, the title field is shown as empty (in edit mode). That is not what happened before. Before the title field contained the text in edit mode as well.

The body seems fine, i.e. as expected with subject line and body joined by newline. (I would have them joined with _two_ newlines, but that is less relevant.)

Can you replicate this behavior?

spenserblack commented on src/main.ts:
```diff
@@ -48,7 +56,7 @@ async function run(): Promise<void> {
     owner,
     repo,
     tag_name: rawTag,
-    name,
+    title,
```
> now I am getting a strange result, where the title shown for the release is `tag: subject` (e.g. `v1.2.3: this is text`.
> 
> What is even stranger though is when I go to manually edit the release, the title field is shown as empty (in edit mode)

The release title is not in the request body when it's posted. GH defaults to that when the title is excluded AFAIK.

The issue is here.
```typescript
const foo = makeFoo();

// In JS/TS, these are equivalent:
{ foo }

{ foo: foo }
```

[The API expects a value in the body called `name`](https://docs.github.com/en/rest/releases/releases?apiVersion=2022-11-28#create-a-release), but this change sends `title` instead.

spenserblack replied:
Thanks, TypeScript, for catching this invalid object (sarcasm) :laughing:

daniil-berg replied:
Ah, see there is my great JS experience. Thanks for the catch. Hope my fix was correct. If you don't mind re-building it.

spenserblack replied:
https://github.com/spenserblack/actions-tag-to-release/pull/114/commits/f99d1d3709e207a602c00b01980bc78bb123079a

daniil-berg commented:
Awesome. Seems to be working as expected. Thanks a lot for taking the time. I expect I'll be using this action a lot. :grinning:
